core/src/main/java/org/apache/calcite/rel/rules/LoptOptimizeJoinRule.java

```diff
@@ -2064,7 +2064,7 @@ private static boolean areSelfJoinKeysUnique(RelMetadataQuery mq,
     // are unique.  When removing self-joins, if needed, we'll later add an
     // IS NOT NULL filter on the join keys that are nullable.  Therefore,
     // it's ok if there are nulls in the unique key.
-    return RelMdUtil.areColumnsDefinitelyUniqueWhenNullsFiltered(mq, leftRel,
+    return RelMdUtil.areColumnsDefinitelyUnique(mq, leftRel,
         joinInfo.leftSet());
   }
```

Comment:
`RelMdUtil#areColumnsDefinitelyUniqueWhenNullsFiltered` has different implementation from `RelMdUtil#areColumnsDefinitelyUnique`. The former one would ignore null while the other one not. Why do you say they have same behavoirs?

Reply:
@chunweilei Thanks a lot for your careful review. They are different indeed. 
However, `areColumnsDefinitelyUniqueWhenNullsFiltered` methods could be implemented in a more efficient way.